src/server/hotspot_partition_calculator.cpp

```diff
@@ -98,5 +104,60 @@ void hotspot_partition_calculator::data_analyse()
     }
 }
 
+/*static*/ void
+hotspot_partition_calculator::send_hotkey_detect_request(const std::string &app_name,
+                                                         const int partition_index,
+                                                         const hotkey_detect_type hotkey_type,
+                                                         const hotkey_detect_action action)
+{
+    dsn::apps::hotkey_detect_request request;
+    request.type = (hotkey_type == hotkey_detect_type::WRITE_HOTKEY_DETECT)
+                       ? dsn::apps::hotkey_type::WRITE
+                       : dsn::apps::hotkey_type::READ;
+    request.action = (action == hotkey_detect_action::STOP_HOTKEY_DETECT)
+                         ? dsn::apps::hotkey_detect_action::STOP
+                         : dsn::apps::hotkey_detect_action::START;
+    ddebug_f("{} {} hotkey detection in {}.{}",
+             (action == hotkey_detect_action::STOP_HOTKEY_DETECT) ? "Stop" : "Start",
+             (hotkey_type == hotkey_detect_type::WRITE_HOTKEY_DETECT) ? "write" : "read",
+             app_name,
+             partition_index);
+    dsn::rpc_address meta_server;
+    meta_server.assign_group("meta-servers");
+    std::vector<dsn::rpc_address> meta_servers;
+    replica_helper::load_meta_servers(meta_servers);
+    for (const auto &address : meta_servers) {
+        meta_server.group_address()->add(address);
+    }
+    auto cluster_name = dsn::replication::get_current_cluster_name();
+    auto resolver = partition_resolver::get_resolver(cluster_name, meta_servers, app_name.c_str());
```

Comment:
You repeat this process in every single call. You should wrap it with a class:

```
class table_hotkey_detect_client {
public:
  void call(const std::string &app_name,
            const int partition_index,
            const hotkey_detect_type hotkey_type,
            const hotkey_detect_action action);

private:
  partition_resolver_ptr _resolver;
}; 
```

Reply:
It's fine to do it in every single call because this function is called in very low frequency, only we need to start/stop hotkey detection on the server we call it.
A class for this simple function is too heavy, I prefer to make it as a static function to use.